LFRic code tutorial with hands-on examples (for #959)

TeranIvy opened the pull request:
This PR for is for the "building LFRic" part of the ESiWACE2 DSL workshop. It uses the LFRic infrastructure from the full LFRic example PR #954.

There are 3 examples with the algorithm and kernel code that needs to be completed. There are solutions for each example that can be built and run. I also modified the `Makefile`s in Joerg's `full_example` and `full_example_netcdf` LFRic examples to remove compiled infrastructure objects and libraries when running `make clean` (something I found useful but forgot to ask for in reviewing his PR).

There's a top-level `Makefile` in `tutorial/practicals/LFRic/building_code` directory that runs `make` on all solutions as a checking mechanism. Unfortunately, I don't think it can be done in Travis due to `gfortran` requirement.

MatthewHambley commented on tutorial/LFRic/README.md:
```diff
@@ -0,0 +1,90 @@
+  https://www.virtualbox.org/)).
+
+* Fortran compiler supporting [Fortran 2003 standard.](
+  https://gcc.gnu.org/wiki/GFortranStandards#Fortran_2003)
```
May require some F2008 features. In particular, when were `submodules` introduced?

TeranIvy replied:
Mentioned briefly with the appropriate links and clarified that F2008 this is not required for the examples in this tutorial.

MatthewHambley commented on tutorial/LFRic/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,202 @@
+Unlike [kernels](LFRic_kernel.md), there is no inheritance from an
+abstract base type. Similar to kernels, however, the general
+encapsulation principle of keeping data and procedures `private`
+unless they are called by other objects still holds.
```
This really just good general practice but it may not add anything to point that out.

TeranIvy replied:
Thought so :)

MatthewHambley commented on tutorial/LFRic/background/LFRic_algorithm.md:
````diff
@@ -0,0 +1,202 @@
+`Makefile`s, see e.g. this code from [`../example1/Makefile`](
+../example1/Makefile)
+
+```bash
````
Shouldn't this be for a Makefile rather than BASH script?

TeranIvy replied:
Yep, good point! Changed.

MatthewHambley commented on tutorial/LFRic/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,206 @@
+algorithm layer and accesses the object data by dereferencing the object
+proxies. The data and other required information is then passed to
+kernels as illustrated in the [*Kernel subroutine* section](
+LFRic_kernel.md#kernel-subroutine).
```
This relates back to a long standing issue I've had. The things we call "proxies" aren't really that in the normal usage of the term. They are much more like "accessors". I don't know what you want to do about this because they are called proxies in the source.

TeranIvy replied:
Good point. I rephrased this as "... PSyclone generates calls to accessor classes of the LFRic data objects passed from the algorithm layer, referred to as *proxies*, and accesses the object data by dereferencing the object proxies." Also rephrased in the algorithm section.

MatthewHambley commented on tutorial/LFRic/background/LFRic_PSy.md:
````diff
@@ -0,0 +1,206 @@
+the proxy data
+
+```fortran
+      INTEGER(KIND=i_def) undf_aspc1_field_out_w0, undf_aspc1_field_out_w3, &
````
I just noticed you're using the common practice of capitalising Fortran keywords. This potentially makes the document more friendly to old-school Fortran programmers but is contrary to actual LFRic policy. Something to consider.

TeranIvy replied:
This is something that comes from PSyclone so I quoted it as it is. I added a note that according to LFRic coding standards we don't capitalise Fortran keyword in the LFRic code.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,241 @@
+
+  !-----------------------------------------------------------------------------
+  ! Public types
+  !-----------------------------------------------------------------------------
```
Could we consider something else here. This comment is really unecessary and unhelpful. Clearly they are public types.

TeranIvy replied:
Changed to
```
  !-----------------------------------------------------------------------------
  ! The type declaration for the kernel. Contains the metadata needed by
  ! the PSy layer.
  !-----------------------------------------------------------------------------
```
here and in all examples.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,241 @@
+
+  !-----------------------------------------------------------------------------
+  ! Contained functions/subroutines
+  !-----------------------------------------------------------------------------
```
Likewise another completely redundant comment.

TeranIvy replied:
Yep, removed here and in all examples.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,241 @@
+  contains
+
+  subroutine setval_field_w0_code()
+
```
Maybe an ellipsis here to mark missing code?

TeranIvy replied:
Good point, added.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
````diff
@@ -0,0 +1,241 @@
+end module setval_field_w0_kernel_mod
+```
+
+As can be seen from the code above, the *LFRic kernels are objects, defined
````
Technically classes.

TeranIvy replied:
Good point, changed.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,241 @@
+  update the field with.
+
+The `GH_` prefix comes from "GungHo", the name of the dynamical core used in
+the LFRic model. Please refer to the documentation for more information on
```
With the Dynamo interface and "GH" prefixes this document really highlights how badly we need to properly rename everything. But you are documenting what is, not what should be.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
````diff
@@ -0,0 +1,241 @@
+           arg_type(GH_FIELD,  GH_REAL, GH_INC, W0), &
+           arg_type(GH_SCALAR, GH_REAL, GH_READ)     &
+           /)
+  ```
````
I understand why you want to include this but is it going to lead to confusion in the reader? Might it be better to leave it out and change it when the implementation changes? Maybe an appendix instead?

TeranIvy replied:
I pushed it to the appendix.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,241 @@
+* `use argument_mod,      only: arg_type, ...` looks for the LFRic
+  identifiers of PSyclone metadata for LFRic objects (e.g. fields);
+* `use fs_continuity_mod, only: ...` looks for the `integer` identifier of
+  the required function space (here `W0`);
```
Although they are implemented as integers that isn't what they really are. They are enumerators, part of an enumerated type.

TeranIvy replied:
I changed this to "... looks for the enumerator identifier (implemente as `integer`) ..."

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,241 @@
+LFRic: everything is `private` unless it needs to be accessed outside the
+object. As a rule, object data are `private` as illustrated in the
+`setval_field_w0_kernel_type` definition above. Object procedures can be
+`private` (if used only by the parent object) or `public`ly available
```
See previous comment about parent object and visibility.

TeranIvy replied:
Again, I don't know how the parent came into this. Removed.

MatthewHambley commented on tutorial/LFRic/background/LFRic_kernel.md:
````diff
@@ -0,0 +1,241 @@
+    field_1_w0( map_w0(df) + k ) = rscalar_2
+  end do
+end do
+```
````
Looping over k in the kernel is an optimisation which assumes that the data is k-first. In the future we may move this looping up to the PSy layer to better handle i-first fields. Should this be noted or omitted as we are still quite far from actually implementing that.

TeranIvy replied:
This was also raised by PSyclons :) Added a note to clarify.

MatthewHambley commented on tutorial/LFRic/example1/Makefile:
```diff
@@ -0,0 +1,101 @@
+
+# Compiler settings
+F90 ?= gfortran
+F90FLAGS ?= -Wall -g -fcheck=bound
```
Do you want to add `-Werror` to promote all warnings to errors? (See also other Makefiles)

TeranIvy replied:
Not really, as the LFRic infrastructure is not quite clear of unused variables and such.

MatthewHambley commented on tutorial/LFRic/example1/Makefile:
```diff
@@ -0,0 +1,101 @@
+PSYCLONE_CMD = -nodm -l all
+
+# Executable
+EXEC = example1.out
```
Often the Windows convention of `.exe` is used for executables. Or the Posix convention of no extension at all.

TeranIvy replied:
Adding `exe` was done so I can have generic `.gitignore` for all examples. However, I went for Posix convention and just specified paths to executables in local `.gitignore` files.

MatthewHambley commented on tutorial/LFRic/example1/Makefile:
```diff
@@ -0,0 +1,101 @@
+	-opsy $*_psy.f90 -oalg $*.f90 $<
+
+clean:
+	rm -f *.o *.mod $(LFRIC_PATH)/*.o $(LFRIC_PATH)/*.mod $(LFRIC_PATH)/*.a \
```
If any of these files don't exist `rm` may return an error causing the Makefile to fail. You can prepend the command with `-` to have Make ignore the error. i.e. `-rm ...`

TeranIvy replied:
`rm` without `-` actually works fine when a file does not exist, so I left them as they are.

MatthewHambley commented on tutorial/LFRic/example2/README.md:
```diff
@@ -0,0 +1,109 @@
+* The current built-ins are defined for `real`-valued field data.
+  Built-ins for `integer`-valued fields will be introduced as part
+  of the support for the field data of other intrinsic types.
+---
```
See previous comment.

TeranIvy replied:
As above, `rm` without `-` actually works fine so I left them as they are.

MatthewHambley reviewed (changes requested):
I notice the workshop examples and what I take to be Jeurg's stripped down LFRic are both in this changset. Is that intentionall?

TeranIvy commented:
> 
> 
> I notice the workshop examples and what I take to be Jeurg's stripped down LFRic are both in this changset. Is that intentionall?

@MatthewHambley , I merged Joerg's branch to mine so I can work until his branch is meged to the master. I'll mark this PR as ready for review after PR #954 is merged and my branch updated with master.

TeranIvy commented on tutorial/LFRic/background/LFRic_intro.md:
```diff
@@ -0,0 +1,140 @@
+* `bin` - [Rose-based](
+   https://www.metoffice.gov.uk/research/approach/modelling-systems/rose) executables;
+
+* `extra` - Utilities (e.g. job submission scripts);
```
Thanks for the info, added a note.

TeranIvy commented on tutorial/LFRic/background/LFRic_intro.md:
```diff
@@ -0,0 +1,140 @@
+compatibility table is give in this [LFRic wiki (requires login)](
+https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicTechnical/VersionsCompatibility).
+
+## LFRic code structure
```
I added an explanation that some of the code used in the Examples is from LFRic repository but it does not depend on it.

TeranIvy commented on tutorial/LFRic/background/LFRic_intro.md:
```diff
@@ -0,0 +1,140 @@
+each is divided into the following layers:
+
+* **Driver** - Set up and control of a model run (e.g. domain, partition,
+  science configuration, checkpoint);
```
Added a brief clarification, thanks.

TeranIvy commented on tutorial/LFRic/background/LFRic_intro.md:
```diff
@@ -0,0 +1,140 @@
+* **Driver** - Set up and control of a model run (e.g. domain, partition,
+  science configuration, checkpoint);
+* [**Algorithm**](LFRic_algorithm.md) - Science code performing high-level
+  mathematical operations (e.g. time-stepping) on LFRic objects;
```
Clarified that in the algorithm section, however it is worh emphasising it here.

TeranIvy commented on tutorial/LFRic/background/LFRic_intro.md:
```diff
@@ -0,0 +1,140 @@
+  repository as `psykal_lite_*_mod.(F)f90` files);
+* [**Kernel**](LFRic_kernel.md) - Science code performing low-level
+  mathematical operations (e.g. matrix assembly, calculating gradient)
+  on the data contained in LFRic objects.
```
Actually, mainly field and operator objects and sometimes scalars.

TeranIvy commented on tutorial/LFRic/background/LFRic_intro.md:
```diff
@@ -0,0 +1,140 @@
+*Algorithms* (natural science performing operations on whole fields).
+
+All of this is supported by the **LFRic infrastructure**, a collection
+of **objects** that encapsulate data and functionality representing the
```
Changed.

TeranIvy commented on tutorial/LFRic/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,202 @@
+they will be referred to as "LFRic data objects".
+
+The important thing to note here is that algorithms (as well as other
+high-level code as drivers) operate on [full objects](
```
Thanks, added.

TeranIvy commented on tutorial/LFRic/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,202 @@
+https://psyclone.readthedocs.io/en/stable/algorithm_layer.html) but
+**do not operate directly on object data** (or, in OO terminology, *must
+not break encapsulation*). The data is accessed in the [PSy layer](
+LFRic_PSy.md) via the required object proxies.
```
Since this enters philosophical domain I left it as it is 😄

TeranIvy commented on tutorial/LFRic/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,202 @@
+`infrastructure` directory of the LFRic repository.
+
+The **object data is private**, whilst the methods (Fortran procedures)
+can be `private` (if used only by the parent object) or `public`ly
```
Tjhanks, changed to "(if used only by the object)" - don't know where the parent came from...

hiker commented on tutorial/LFRic/example3/example3_driver.f90:
```diff
@@ -0,0 +1,220 @@
+  ! Set partitioner parameters (planar mesh on a single process)
+  !-----------------------------------------------------------------------------
+  max_stencil_depth = 0
+  local_rank = 0
```
This line must be moved to before the read_configuration call in line 125, otherwise local_rank is undefined when the configuration is read.

TeranIvy replied:
Corrected, thanks for spotting! I wondered why it didn't fall apart and printing the value revealed that `gfortran` initialised it to 0. Compilers can be too helpful :worried:

hiker replied:
My gfortran didn't - that's how I noticed  :)

TeranIvy commented:
Tests pass so over to @arporter for review.

arporter reviewed (changes requested):
I'm pausing my review because I've already hit 100 comments (and don't want to lose them!) and also because, having looked at the first tutorial I think it needs a lot more hand holding.
You've been very, very thorough but I think you are perhaps expecting too much of attendees!
Even as someone who understands PSyclone and LFRic, I struggled to understand the purpose of each part - i.e. what I was working towards and how I would know when I got there.
It was also not clear to me when the student was supposed to look at the 'background' material. I think much/all of that should be moved into the hands-on, step-through description so that at each point the user knows what they're looking at.
Maybe we should just iterate on the part of the tutorial that I've reviewed so far and then when we're both happy, you can apply the same pattern to the remainder of it?

TeranIvy commented:
Thanks for the review, @arporter. Before I start working on this, I'll outline what I understood should be done here so I don't go down the wrong route:

- Most of the Kernel section with the specifics on metadata, arguments and loops needs to go in the README of example 1 (I guess tutorial 1),
- Algorithm part can be split into two parts: one that explains invokes to kernels (that goes into exercise 1) and the other that explains inokes to built-ins code (that goes into exercise 2)
- Similarly, PSy-layer section can be split into two parts: one that deals with the code for generated kernel calls (that goes into exercise 1) and the other that deals with the built-ins code (that goes into exercise 2),
- The general overviews of what each layer does (e.g. kernels operate on data, etc) can then go into a single README file as a reference. How about just putting them into LFRic_intro?

Is this roughly what you had in mind?

arporter commented:
Hi @TeranIvy, yes, that's pretty much it. Are you going to introduce some of the background material in your slides?

TeranIvy commented on examples/lfric/full_example/Makefile:
```diff
@@ -77,4 +78,5 @@ main_alg.f90: main_psy.f90
 	psyclone --config ../../../config/psyclone.cfg -opsy $*_psy.f90 -oalg $*_alg.f90 $<
 
 clean:
```
Thanks for the suggestion, done.

TeranIvy commented on examples/lfric/full_example_netcdf/Makefile:
```diff
@@ -77,4 +78,5 @@ main_alg.f90: main_psy.f90
 
 clean:
-	rm -f *.o *.mod  $(EXEC) main_alg.f90 main_psy.f90
+	rm -f *.o *.mod  $(EXEC) main_alg.f90 main_psy.f90 \
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/domain_size_config_mod.f90:
```diff
@@ -9,7 +9,7 @@
 ! BSD 3-Clause License
 !
-! Copyright (c) 2020, Science and Technology Facilities Council
+! Modifications opyright (c) 2020, Science and Technology Facilities Council
```
Oops, corrected.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/field_mod.F90:
```diff
@@ -719,7 +721,11 @@ subroutine log_minmax( self, log_level, label )
     write( log_scratch_space, '( A, A, A, 2E16.8 )' ) &
          "Min/max ", trim( label ),                   &
-         " = ", fmin%get_min(), fmax%get_max()
+    ! Functions get_min() and get_max() rely on MPI comms so
```
@hiker's example calls `log_minmax`, however it fails to run due to the call to `mpi_mod` to enable MPI comms for calculations of the global minima and maxima. The `mpi_mod` will compile but most of the machinery is switched off so to not call MPI comms (I guess Joerg did this to mimimize changes to the code and still not allow MPI).
Since MPI is not supported, the minimum and maximum of a field is not calculated for each MPI process but just once from the field data. Does this make sense?

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/domain_size_config_mod.f90:
```diff
@@ -121,8 +121,6 @@ subroutine read_namelist( file_unit, local_rank )
     buffer_real_r_def(3) = planar_domain_min_x
     buffer_real_r_def(4) = planar_domain_min_y
```
Reading in namelists in LFRic also works with MPI broadcast. Here it was compiled but was not run as this stripped-down infrastructure library does not support MPI. I disabled the MPI broadcast call because I actually need to run this config file to read a namelist in one of my tutorial examples.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+# Tutorial on building LFRic code
+
+This tutorial gives an overview of building LFRic code, with PSyclone
+as an integral part of this process. This repository is structured
```
Ok, done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+
+This tutorial gives an overview of building LFRic code, with PSyclone
+as an integral part of this process. This repository is structured
+into four subdirectories, one with the
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+## LFRic code support
+
+All the examples here use a pared-down version of the LFRic infrastructure
+stored in `src/psyclone/tests/test_files/dynamo0p3/infrastructure`
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+
+All the examples here use a pared-down version of the LFRic infrastructure
+stored in `src/psyclone/tests/test_files/dynamo0p3/infrastructure`
+directory of PSyclone repository. For more information on LFRic code
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+background/LFRic_intro.md#lfric-repository-structure) of the
+[introduction to LFRic](background/LFRic_intro.md).
+
+### Implementation of PSyclone in examples
```
It could. Changed.

TeranIvy replied:
Since I am pointing to @rupertford's tutorials on distributed and shared memory support I changed the title to "Parallel code support".

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+https://psyclone.readthedocs.io/en/stable/transformations.html) are
+applied here so the code is generated and run in serial.
+
+## Software environment
```
Moved - I guess this will be topped up with the some more info from the GoogleDoc?

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+  https://psyclone.readthedocs.io/en/stable/getting_going.html)
+  The simplest way is to install PSyclone into the Python 3 environment
+  that will be used in building LFRic. Alternatively, it can be installed in
+  `user` mode or to another location. Note, the latter requires explicit
```
Ok, I removed this. Since this document is not purely about LFRic any more, I also removed the reference to Python 3.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+* [NetCDF library](https://www.unidata.ucar.edu/software/netcdf/),
+  together with its Fortran and C++ bindings, is required for reading in LFRic
+  mesh files written in
+  [UGRID format](https://ugrid-conventions.github.io/ugrid-conventions/).
```
I plundered the workshop requirements GoogleDocs for NetCDF in the end, I hope you don't mind :smiley:

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+  of requirements, is described [here.](
+  https://www.unidata.ucar.edu/software/netcdf/docs/getting_and_building_netcdf.html)
+  The above-listed recent ersions of NetCDF packages were built with
+  [HDF5](https://www.hdfgroup.org/solutions/hdf5) 1.10.6 as one of the
```
If NetCDF is built via the package manager then they will be included. If not they will be required so I noted it here.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+  tutorial either [Python 2](https://www.python.org/download/releases/2.0/)
+  or [Python 3](https://www.python.org/download/releases/3.0/) environments
+  are fine (plotting scripts require `numpy` and `matplotlib` packages in
+  either case). Running the entire LFRic code, however, requires Python 3
```
As this is now general-purpose requirements document I removed this line.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+compatibility table is give in this [LFRic wiki (requires login)](
+https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicTechnical/VersionsCompatibility).
+
+## LFRic repository structure
```
Removed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+
+---
+
+### Languages and software requirements
```
Removed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+
+## Separation of concerns
+
+Applications in LFRic directories build on the principle of *separation
```
Yep, changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+* [**Algorithm**](LFRic_algorithm.md) - Science code performing high-level
+  mathematical operations (e.g. solver preconditioning) on LFRic field and
+  operator objects;
+* [**PSy**](LFRic_PSy.md) - Parallel code either generated by PSyclone (not
```
There is no PSyKAl-lite code in this tutorial so I left only the code generation part here.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+  repository as `psykal_lite_*_mod.(F)f90` files);
+* [**Kernel**](LFRic_kernel.md) - Science code performing low-level
+  mathematical operations (e.g. matrix assembly, calculating gradient)
+  on the data held in LFRic field and operator objects.
```
Good point, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+  mathematical operations (e.g. matrix assembly, calculating gradient)
+  on the data held in LFRic field and operator objects.
+
+This above illustrated separation of concerns is called **PSyKAl**, from
```
Ok, done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+  on the data held in LFRic field and operator objects.
+
+This above illustrated separation of concerns is called **PSyKAl**, from
+*Parallel Systems* (computational science applying optimisations),
```
I think we usually talk in plural in LFRic, but a look into PSyclone ReadTheDocs reveals singular, so I changed them to be consistent with the PSyclone documentation.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+**do not operate directly on object data** (or, in OO terminology, *must
+not break encapsulation*). The data is accessed in the [PSy layer](
+LFRic_PSy.md) via the required object accessor class, referred to as
+*proxy* in the LFRic code.
```
Bleeping articles... Done, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+examples, [`simple_kernels_alg_mod.x90`](
+../1_simple_kernels/solutions/part1/simple_kernels_alg_mod.x90) and
+[`builtins_alg_mod.x90`](../2_built_ins/solutions/builtins_alg_mod.x90),
+to illustrate provisional structure of LFRic algorithms with `invoke` calls
```
Changed. Note that I'm addressing inline comments now before I chop up these bits to tasks 1 and 2 where they need to go.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+to illustrate provisional structure of LFRic algorithms with `invoke` calls
+to kernels and built-ins.
+
+Both examples create fields on different function spaces on basis of the
```
Done (also prior to moving the relevant parts of this docs to exercises 1 and 2).

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+[PSyclone built-ins](
+https://psyclone.readthedocs.io/en/stable/dynamo0p3.html#built-ins).
+
+The `invoke` calls to kernels are really calls to the *kernel objects*
```
Yes, I forgot a lot of background here. Rephrased prior to refactoring this document.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+
+## Processing of algorithm code in LFRic
+
+As can be seen from examples, algorithm code in LFRic has `x90` extensions
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/Makefile:
```diff
@@ -0,0 +1,46 @@
+# Top-level Makefile to build all solutions of the LFRic hands-on tutorials on
+# building code.
+
+.PHONY: default clean
```
I have updated the `test` target to just generated alg and PSy source. `default` is now `build` (works with either `make` or `make build`).

TeranIvy commented on tutorial/practicals/LFRic/building_code/Makefile:
```diff
@@ -0,0 +1,46 @@
+# Author: I. Kavcic, Met Office
+
+# Top-level Makefile to build all solutions of the LFRic hands-on tutorials on
+# building code.
```
Yep, added `test` target for just PSyclone in all Makefiles. Also updated `.travis.yml` and `.github/workflows/python-package.yml`.

TeranIvy replied:
P.S. I listed the `make` targets in the main `README.md` and advised the use of `make test` instead of full `make build` whilst completing the kernel and algorithm code in order to utilise PSyclone checks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+# LFRic PSy
```
Added "layer" but I don't think articles are usally part of a title.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+[*PSy layer in LFRic in PSyclone documentation*](
+https://psyclone.readthedocs.io/en/stable/dynamo0p3.html#psy-layer)
+
+## What PSy layer does
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+
+## What PSy layer does
+
+As said in the [*What algorithms do* section](
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+As said in the [*What algorithms do* section](
+LFRic_algorithm.md#what-algorithms-do), LFRic algorithms perform
+mathematical operations on LFRic data objects and pass them as
+arguments to kernel and built-in calls in `invoke`s. As said in the
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+
+PSyclone also generates code for distributed and shared memory
+support, however this is not utilised in this tutorial as explained
+in the main [README document](../README.md). The below examples of
```
Done, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+## Generated code to support kernel calls
+
+As illustrated in the [*`invoke` calls* section](
+LFRic_algorithm.md#invoke-calls) of algorithm introduction,
```
Thanks for the suggestion, done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
````diff
@@ -0,0 +1,223 @@
+    call invoke( setval_field_w0_kernel_type(field_w0, scalar_w0) )
+```
+
+translates to this PSy-layer subroutine call
````
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+provided in the [first example, solutions of Part 1](
+../1_simple_kernels/solutions/part1) (generated by running `genkernelstub`).
+
+The above kernel call code call illustrates how `iterates_over = CELLS`
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+
+---
+
+## Generated code to support built-in calls
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+## Generated code to support built-in calls
+
+As illustrated in the [*`invoke` calls* section](
+LFRic_algorithm.md#invoke-calls) of algorithm introduction,
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+is located in the `builtins_alg_mod_psy.f90` file (also not kept in the
+repository).
+
+As for a kernel call there are, amongst others, generated declarations and
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
````diff
@@ -0,0 +1,223 @@
+As for a kernel call there are, amongst others, generated declarations and
+calls to the `field_out_w0`, `field1_in_w0` and `field2_in_w0` proxies
+
+```fortran
````
I will leave it as it is as this part will be going to tutorial 2 on built-ins instead of staying here, so it will be the illustration of what happens with built-in calls in the PSy layer. I added the key point, though.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+
+Unlike the PSy-layer kernel code calls, the generated code is a direct
+representation of the mathematical operation that a built-in performs
+(in this case, `field_out_w0 = field1_out_w0 + field2_out_w0`)
```
Rephrased, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+# LFRic kernel
+
+[*Kernel layer in LFRic in PSyclone documentation*](
```
Done here and in Algorithm and PSy sections.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+## Kernel structure
+
+In this section we use [`setval_field_w0_kernel_mod.f90`](
```
Done before moving this with the first tutorial.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+[PSy layer](LFRic_PSy.md).
+
+Scalar values are passed to a kernel from the
+[algorithm layer](LFRic_algorithm.md) as they are. In case of fields
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+Scalar values are passed to a kernel from the
+[algorithm layer](LFRic_algorithm.md) as they are. In case of fields
+and operators, however, LFRic kernels operate on a subset of degrees
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+Scalar values are passed to a kernel from the
+[algorithm layer](LFRic_algorithm.md) as they are. In case of fields
+and operators, however, LFRic kernels operate on a subset of degrees
+of freedom (DoFs) of these objects. Such a subset occupies a portion
```
Thanks for the suggestion, done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+[algorithm layer](LFRic_algorithm.md) as they are. In case of fields
+and operators, however, LFRic kernels operate on a subset of degrees
+of freedom (DoFs) of these objects. Such a subset occupies a portion
+of the computational domain (e.g. an entire mesh or a partition of a
```
I think it makes sense to do that to have link with distributed memory. I am happy to link it to Rupert's examples where appropriate.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+Looking at the `setval_field_w0_kernel_type` above, there are two main
+parts of an LFRic kernel:
+* Metadata that tells PSyclone how to operate on a kernel;
```
Changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+### Kernel metadata
+
+The PSyclone metadata are stored as the *private* data within the definition
```
Changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+* A field argument (`GH_FIELD`) with the degrees of freedom (DoFs) on `W0`
+  function space that is being updated in this kernel (`GH_INC` access for
+  updating fields on continuous function spaces);
+* A `real`-valued scalar (`GH_REAL`) whose value is read (`GH_READ`) to
```
Good point, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+from the [PSy layer](LFRic_PSy.md). A user-defined kernel in LFRic
+is called from a PSy-layer loop over each cell in the horizontal domain,
+hence the metadata identifier for this way of looping, `CELLS`. This means
+that a kernel operates on a one-cell-wide vertical column of cells.
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+#### Argument list and declarations
+
+Running PSyclone
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+Running PSyclone
+[kernel stub generator](https://psyclone.readthedocs.io/en/stable/stub_gen.html)
+on `setval_field_w0_kernel_mod.f90` produces the argument list and
+declarations for the `setval_field_w0_code` enclosed in a module:
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+The generated subroutine body with the argument list and declarations can
+be used to complete the kernel stub above (note that the generated
+`USE constants_mod` statement is redundant as per LFRic convention it is
```
You're right, it's not redundant. It's just at the wrong level per LFRic convention. Clarified.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+already declared at the module level).
+
+`rscalar_2` denotes the read-only scalar argument passed from the
+[LFRic algorithm layer](LFRic_algorithm.md) as is. As said
```
Thanks for the suggestion, done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+  column that the kernel operates on);
+* `field_1_w0` is the data stored in an LFRic field object;
+* `ndf_w0` is the number of all DoFs (owned and annexed) for the field
+  argument defined on `W0` function space;
```
There are no articles in my mother tongue and we're getting on fine without them :smiley: Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+* `map_w0` is the DoF-map for the cell at the base of the column for
+  the `W0` function space that this field is defined on (the map
+  contains local IDs of DoFs in a partition).
+
```
Good point and thanks for the suggestion, done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+The order of kernel arguments roughly corresponds to the order of
+[kernel metadata descriptors](#kernel-metadata). The convention when
+ordering arguments is to put arguments that are written to first.
```
Thanks for the suggestion, added appropriate links and rephrased.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+---
+**NOTE**
+
+Looping over `k` in the kernel is an optimisation which assumes that the
```
I think it is important as people may wonder why and we are working towards other ways of looping. I rephrased the note to say that a) this is a more natural way of writing FEM numerics for scientists as you suggested and b) it is an optimisation from the computational science viewpoint and we are looking into other options.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+[introduction to LFRic](background) (including the overview of LFRic
+code structure and how PSyclone is used in LFRic) and three examples.
+
+## Examples
```
Restructured and made clear that they need to be worked through in order.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+This tutorial gives an overview of building LFRic code, with PSyclone
+as an integral part of this process. This repository is structured
+into four subdirectories, one with the
+[introduction to LFRic](background) (including the overview of LFRic
```
Not necessarily, I wrote that the recommended background sections are highlighted and linked in each tutorial.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+
+### Part 1
+
+Use PSyclone kernel stub generator to create argument list and
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+### Part 1
+
+Use PSyclone kernel stub generator to create argument list and
+declarations for two kernels, one that assigns a value to a field on
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+continuous `W0` function space and another on a discontinuous `W3`
+function space. Modify the supplied algorithm
+[`simple_kernels_alg_mod.x90`](simple_kernels_alg_mod.x90) to call
+these kernels from.
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+The stub for the first kernel with the required metadata can be found
+in the file [`setval_field_w0_kernel_mod.f90`](
+setval_field_w0_kernel_mod.f90). Declarations and argument
+list code can be created by running PSyclone kernel stub generator:
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_kernel.md:
```diff
@@ -0,0 +1,258 @@
+
+In this section we use [`setval_field_w0_kernel_mod.f90`](
+../1_simple_kernels/setval_field_w0_kernel_mod.f90) kernel stub
+from the [simple kernels example](../1_simple_kernels) to
```
Moved the `LFRic_kernel.md` document to `1_simple_kernels` and renamed it to  `LFRic_kernel_structure.md`. The general overview of what kernels do is now in `background/LFRic_structure.md` and  `LFRic_kernel_structure.md` gives the specific details on kernel argument list and loop structure.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+Use PSyclone kernel stub generator to create argument list and
+declarations for two kernels, one that assigns a value to a field on
+continuous `W0` function space and another on a discontinuous `W3`
+function space. Modify the supplied algorithm
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+
+The stub for the first kernel with the required metadata can be found
+in the file [`setval_field_w0_kernel_mod.f90`](
+setval_field_w0_kernel_mod.f90). Declarations and argument
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,127 @@
+  (does not need to be modified). Run `make` to build the completed
+  example and `./simple_kernels` to run it;
+* [`simple_kernels_driver.f90`](simple_kernels_driver.f90) - an example
+  of LFRic-like main program that creates the required LFRic objects and
```
Done.

TeranIvy commented on tutorial/practicals/LFRic/building_code/README.md:
```diff
@@ -0,0 +1,96 @@
+https://www.dkrz.de/redmine/projects/yaxt) in LFRic). Also, none of
+the [PSyclone transformations](
+https://psyclone.readthedocs.io/en/stable/transformations.html) are
+applied here so the code is generated and run in serial.
```
Added links to both distributed and shared memory tutorials.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_intro.md:
```diff
@@ -0,0 +1,155 @@
+spaces, fields, operators, etc. The schematic of the overall PSyKAl
+structure is shown in the figure below (courtesy of the UKMO LFRic team).
+
+![PSyKAl Separation of concerns in LFRic](separation_concerns.png)
```
I don't know, it's been floating around for ever. I left it as it is for now as I haven't heard back about authorship.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+manipulate data objects via `invoke` calls to kernels and [built-ins](
+https://psyclone.readthedocs.io/en/stable/dynamo0p3.html#built-ins).
+
+In this section we use the completed algorithms from the solutions of
```
Distributed the content of this section into the relevant tutorials and other Markdown files.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_algorithm.md:
```diff
@@ -0,0 +1,204 @@
+
+The `invoke` calls to kernels are really calls to the *kernel objects*
+(defined as derived types, as explained in the [*LFRic kernel* section](
+LFRic_kernel.md). An example of such call can be found in the
```
Distributed the content of this section into the relevant tutorials and other Markdown files.

TeranIvy commented on tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md:
```diff
@@ -0,0 +1,223 @@
+[*Kernel subroutine* section](LFRic_kernel.md#kernel-subroutine).
+
+PSyclone also generates code for distributed and shared memory
+support, however this is not utilised in this tutorial as explained
```
Pointed to tutorials. Note that the contents of this file are now distributed amongst tutorials 1 and 2 and the background information.

TeranIvy commented:
Right, I have restrucured the documentation and simplifiefd code in the tutorials 1 and 3. I also provided step-by-step instructions as suggested. Ready for another read-through from @arporter.

hiker commented on tutorial/practicals/REQUIREMENTS.md:
```diff
@@ -0,0 +1,44 @@
+  are fine (plotting scripts require `numpy` and `matplotlib` packages in
+  either case).
+
+* Latest PSyclone release (currently 1.9.0). For installation please refer to
```
I just noticed this. Should this be the current git release, not the latest official one?

arporter replied:
Ah! We have a bit of a clash here as I've added a `tutorial/practicals/README.md` in my 933_ branch. We definitely shouldn't be using the word 'release' anywhere :-)

TeranIvy replied:
Shall I delete this document then, @arporter?

arporter replied:
Yes, if you wouldn't mind - I think that's the simplest solution.

TeranIvy replied:
Deleted

hiker commented on tutorial/practicals/LFRic/building_code/3_time_evolution/solutions/prop_perturbation_kernel_mod.f90:
```diff
@@ -0,0 +1,140 @@
+    integer(kind=i_def), intent(in) :: ndf_w3
+    integer(kind=i_def), intent(in) :: undf_w3
+    integer(kind=i_def), intent(in), dimension(ndf_w3) :: map_w3
+    real(kind=r_def), intent(inout), dimension(undf_w3) :: perturbation
```
Perturbation is declared as GH_READWRITE, and intent(inout) - but as far as I can tell it is only written to, not read.

TeranIvy replied:
This should actually be `write` whether `init_perturbation` is `readwrite` as there is a built-in call beforehand. Changed,

arporter commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/README.md:
```diff
@@ -0,0 +1,162 @@
+    LFRic_kernel_structure.md#argument-list-and-declarations) and
+  - [Loops](LFRic_kernel_structure.md#loops) to update LFRic field objects.
+
+The tutorial is further subdivided into two parts in order to learn how to
```
I'd make it clearer that these *are* the two parts of the tutorial and that a user should do each in turn. I keep reading this file and wondering what I'm supposed to do.

TeranIvy replied:
Thanks, I rephrased this and listed the supporting source accordingly.

arporter commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/LFRic_kernel_structure.md:
```diff
@@ -0,0 +1,265 @@
+# LFRic kernel structure
+
+In this part of the tutorial we learn about the structure of an LFRic
```
When do users do this part of the tutorial? Can the material here be inserted into the hands-on steps instead?

TeranIvy replied:
This is supporting material for the first tutorial are using it along the tasks of completing kernels. The READMEs in parts1 and 2 are too long to insert more things, but I rephrased them to clarify this document should be accessible and copied from to complete the tasks.

arporter reviewed (changes requested):
Thanks for re-structuring Iva, I think it's better now.
A couple of the README's need tidying/restructuring slightly to make it clear what steps a user is expected to do.
I'm a bit worried about how long it's going to take people to do. How long were you envisaging giving people for each part?

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
```diff
@@ -0,0 +1,280 @@
+
+You should now have the completed `setval_field_w0_kernel_mod.f90`
+kernel. To check that everything is correct, look into the completed
+kernel in the [Solutions of Part 1 directory](../solutions/part1).
```
Moved.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
```diff
@@ -0,0 +1,280 @@
+
+The new kernel updates a field on a discontinuous function space
+`W3` so you will need to correct the access metadata for the updated
+field from `GH_INC` to `GH_READWRITE` (see [here](
```
Changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
```diff
@@ -0,0 +1,280 @@
+The generated `invoke_0` subroutine code is located in the generated PSy-layer
+file, `simple_kernels_alg_mod_psy.f90`.
+
+There are generated declarations and a calls to the `field_w0` and `field_w3`
```
Thanks, changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
```diff
@@ -0,0 +1,280 @@
+as well as generated declarations and dereferencing assignments to access
+the proxy data required for the full kernel call argument list (see the
+[*Kernel subroutine* section](
+../LFRic_kernel_structure.md#kernel-subroutine))
```
Added.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
````diff
@@ -0,0 +1,280 @@
+      undf_w3 = field_w3_proxy%vspace%get_undf()
+```
+
+Finally, there are generated calls to the tow kernel subroutines,
````
Sorry! Changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
```diff
@@ -0,0 +1,280 @@
+
+with the same argument order as generated by using the `genkernelstub`.
+
+The above kernel calls code also illustrates how the `iterates_over = CELLS`
```
Removed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
````diff
@@ -0,0 +1,237 @@
+will run the kernel stub generator on the renamed kernel
+
+```shell
+genkernelstub setval_field_any_kernel_mod.f90
````
Ok, I reworked this part to start with a kernel stub for `setval_field_any_kernel_mod.f90` and not hassle with copying and renaming `setval_field_w0_kernel_mod.f90`.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
```diff
@@ -0,0 +1,237 @@
+
+You will now need to update the implementation of the kernel loop so
+that it updates the elements of `field_1` to the sum of elements of
+the read-only fields `field_2` and `field_2` to something like
```
Changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
```diff
@@ -0,0 +1,237 @@
+identifier.
+
+The argument list has changed significantly so we need to run the
+kernel stub generator to update the argument list and declarations.
```
Thanks, added the note and a link to Fortran wiki on hoe to break lines (in case people don't know it).

arporter replied:
I don't think you need the link to the wiki - just get them to use the `-l` flag.

TeranIvy replied:
Rephrased, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/README.md:
```diff
@@ -0,0 +1,280 @@
+
+## Step 5: Build and run the code
+
+We will now copy the `simple_kernels_driver.f90` file to this working directory
```
Done and modified.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
```diff
@@ -0,0 +1,237 @@
+../part1/README.md) of this tutorial, so in this tutorial we go directly
+to building and running the code. 
+
+We will copy the `simple_kernels_driver.f90` file to this working directory
```
Done and modified.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
````diff
@@ -0,0 +1,237 @@
+                 ...
+```
+
+It is customary in LFRic to specify the Fortran `kind` of a literal value,
````
You're right, rephrased.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
```diff
@@ -0,0 +1,237 @@
+
+We will now create the `invoke` call to the above generic kernels to
+1. Initialise the output field `field_w0_out` to `0`,
+2. Initialise the input fields `field1_w0_in` and `field2_w0_in` to
```
I specified that we here use -2 and 6 but other values can be used.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
````diff
@@ -0,0 +1,237 @@
+```
+
+We will then create another `invoke` that repeats the calls for the `W3`
+fields, but with the different values of scalar literals for initialising
````
Specified and clarified that other values can be used.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
````diff
@@ -0,0 +1,237 @@
+
+```fortran
+do k = 0, nlayers-1
+  do df = 1, ndf_w0
````
Oh yes. Changed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+[*Driver structure*](#driver-structure) section below outlines the role
+of the driver layer in this tutorial.
+
+### Driver structure
```
Removed and left a reference for the "Driver structure" in Tutorial 1.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+../1_simple_kernels/README.md#driver-structure), so please look there
+for the explanation of the relevant Fortran calls.
+
+### Algorithm structure
```
Removed the section and added the references and the point about the lack of  `use` statements where appropriate as suggested.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+The provided [`builtins_alg_mod.x90`](builtins_alg_mod.x90) source
+defines output and two input fields on function spaces `W0` and `W3`,
+respectively. It outlines the creation of fields on the `W0` function
+space and prints out their `min` and `max` values. The tasks in this
```
Changed, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+
+* Create the output and input fields on the `W3` space using the `W0`
+  fields as a template;
+* Use built-ins to perform the following mathematical operations:
```
I left it as it is as the bullet points below define what operations to perform, in what order and what scalar values to use.

arporter replied:
Thanks, I think that looks better.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+
+This step consists of applying the following mathematical operations:
+
+* Initialising (setting) fields to a scalar value (e.g. `field_out_w0 = 0`),
```
I moved the bullet point list with the specific information to this step, just above this part. Hopefully it is now clear what to do.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+* Initialising (setting) fields to a scalar value (e.g. `field_out_w0 = 0`),
+  described by the [`setval_c` built-in](
+  https://psyclone.readthedocs.io/en/stable/dynamo0p3.html#setval-c);
+* Adding two fields (`field_out_w0 = field1_out_w0 + field2_out_w0`),
```
Oops, changed here and in the code.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+specify the Fortran `kind` of a literal value, e.g. `r_def` for the
+`real`-valued literals in these calls).
+
+Copy the `invoke` stub above to the algorithm source and complete it.
```
Added explanation with the first two steps of the process.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
````diff
@@ -0,0 +1,369 @@
+                 setval_c(field_out_w3, 0.0_r_def),                  &
+```
+
+As in the [Part 2](../1_simple_kernels/part2/README.md) of the simple
````
Added a note (couldn't find the issue number, I'm sure there is one though).

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+
+### Step 3: Use PSyclone to generate algorithm and PSy-layer source
+
+We will first check that the built-in calls in the algorithm code are
```
Re-iterated, thanks.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+
+Looking at the generated algorithm file `builtins_alg_mod.f90` we can
+see that the original `invoke` call to built-ins has become a call
+to a single subroutine with all field arguments passed to the built-ins:
```
Corrected.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+file, `builtins_alg_mod_psy.f90`. We will look at the generated code for
+just the `X_plus_Y` built-in below.
+
+As for a kernel call there are, amongst other things, generated declarations
```
Corrected.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
````diff
@@ -0,0 +1,369 @@
+      ! Initialise number of DoFs for aspc1_field2_in_w0
+      !
+      undf_aspc1_field2_in_w0 = field2_in_w0_proxy%vspace%get_undf()
+```
````
Added - almost literally if you don't mind.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+---
+**NOTE**
+
+As for [LFRic kernels](../1_simple_kernels/LFRic_kernel_structure.md#appendix),
```
Ok, removed.

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/README.md:
```diff
@@ -0,0 +1,369 @@
+space metadata `ANY_SPACE_1` used to define built-ins in PSyclone (see
+the [quick intro to built-ins below](#quick-intro-to-built-ins) below).
+
+### Step 4: Print out min/max values of the fields on the `W3` function space
```
Removed and updated the algorithm code to just print it out.

TeranIvy commented on tutorial/practicals/LFRic/building_code/Makefile:
```diff
@@ -0,0 +1,49 @@
+build test clean:
+	${MAKE} -C 1_simple_kernels/solutions/part1 $@
+	${MAKE} -C 1_simple_kernels/solutions/part2 $@
+	${MAKE} -C 2_built_ins/solutions $@
```
This may have been due to dependency on `KERNEL_OBJ` that was just copied and pasted from other Makefiles. I removed it for the built-ins tutorial. Please let me know if it works now.

arporter replied:
Ah, should I have been using the `test` target? If I do that then everything is fine!

TeranIvy replied:
For me both `make test` and `make build` work for all solutions from this directory. Can you try `make build` from here again and see what happens? More importantly, do both targets from in the `2_built_ins/solutions` directory for you?

arporter replied:
Oh yes, `build` works OK. Maybe we should just change the default target to be either `test` or `build`? (`.DEFAULT_GOAL := build`)

TeranIvy replied:
The default goal is build as it is listed as the first target and this happens when I run `make`. I agree it is good to explicitly specify it so I'll do this in the `Makefile`s.

TeranIvy replied:
In fact, I changed `build` to `compile` to be in line with `examples`. Done for all `Makefile`s here.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+a "perturbation signal" field to the prescribed analytical function
+and the time evolution of that field.
+
+The time evolution means that we need to have a **timestepping loop**.
```
Replaced as appropriate.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+  output routine to write the "model state" consisting of the coordinate and
+  perturbation fields to a file;
+
+* [`gungho_lib`](gungho_lib), a collection of the LFRic infrastructure and
```
It is not needed there but it is here unless I wanted to write essentially the same code as in  LFRic  for what I wanted to do here (e.g. output fields, read namelists and assign coordinates). I chose to add some more LFRic code instead. I clarified that these are additional libraries to what is in the pared-down infrastructure.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+  behaviour and timestepping options;
+
+* `mesh_planar100x100-1000x1000.nc`, an input 2D global planar mesh in the
+  NetCDF-based UGRID format to create the model domain (3D partitioned
```
Thanks, modified.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+  NetCDF-based UGRID format to create the model domain (3D partitioned
+  mesh) from (does not need to be modified). The horizontal limits of
+  the mesh are written down in the *`domain_size`* namelist (must not be
+  modified) in the `configuration.nml` file. Viewing the file requires
```
Dropped.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+where these loops call algorithm subroutines that calculate the
+"model state" for one timestep.
+
+The diagnostic output is produced as a file named `model_state_tstep_<n>.txt`
```
Removed and rephrased.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+
+---
+
+As said above, this driver also shows two other functions that the driver
```
Well, if I need to explicitly say to them to open a file in an editor then it would be a good thing to leave this, too.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+
+As in the previous two tutorials, this algorithm also calls the LFRic field
+class `log_minmax` procedure for quick checks of the the minimum and maximum
+values of the fields after calling the kernels that update them. One of the
```
Ok, just added the calls and updated the materials.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+DoF-wise access to the coordinate fields data and storing the intermediate
+calculation results for the transformed `x` and `y` coordinates and the
+height-varying perturbation amplitude. For instance, the expression
+*(x - x<sub>c</sub> - u\*t<sub>tot</sub>)* can be written in the loop body as
```
Sorry, this was a copy-paste error! It is supposed to be `*(x - x<sub>c</sub>)*`. Corrected.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+[`time_evolution_driver.f90`](time_evolution_driver.f90) as `tstep`.
+Look for the comment that marks the place to calculate the total time
+and complete the `invoke` call in the `time_evolution_alg_step` subroutine,
+`! TO COMPLETE: Propagate the perturbation field in a loop over timesteps`
```
Ok, thanks, rephrased and update the comments.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/gungho_lib/perturbation_bell_config_mod.f90:
```diff
@@ -0,0 +1,209 @@
+!-----------------------------------------------------------------------------
```
They do not need to edit this code in LFRic, though, as it is generated for them. They just need to edit the namelist values.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/plot_xy_slices_ex3.py:
```diff
@@ -0,0 +1,135 @@
+                         "timestep %s" % (level, height, tstep))
+
+    # Show plots on all levels
+    plt.show()
```
Saved as PNG as suggested for each level-timestep combination and added `*.png` target to `make clean`.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+
+### Step 6: Plot the output model state
+
+As said [above](#supporting-source-and-scripts), the supplied Python
```
Saved as PNG as suggested for each level-timestep combination and added `*.png` target to `make clean`. I updated the Step 6 to note this.

arporter replied:
Great.

TeranIvy commented on tutorial/practicals/LFRic/building_code/3_time_evolution/README.md:
```diff
@@ -0,0 +1,441 @@
+
+### Driver structure
+
+[`time_evolution_driver.f90`](time_evolution_driver.f90) follows the order
```
Moved to appendix.

TeranIvy commented:
Addressed the comments. Ready for another review from @arporter.

arporter commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
```diff
@@ -56,34 +56,72 @@ a generic discontinuous function space `ANY_DISCONTINUOUS_SPACE_n`
+
+The argument list and declarations are generated in one line per
+statement each, so in this case of long argument names they may overrun
+the free-form Fortran line-length limit of 132 characters and PSyclone
```
They can just use the `-l` flag to `genkernelstub`.

TeranIvy replied:
Rephrased, thanks.

arporter commented:
I've fixed some minor typos but there are a couple of things that I need you to look at. Once they're done, I shall merge this.

TeranIvy commented:
> I've fixed some minor typos but there are a couple of things that I need you to look at. Once they're done, I shall merge this.

Ok, thanks. I'll also merge the latest master, fix conflicts and double-check that all links work.

TeranIvy commented on tutorial/practicals/LFRic/building_code/1_simple_kernels/part2/README.md:
```diff
@@ -17,36 +21,32 @@ The working directory for this part of the tutorial is
-the completed kernel `setval_field_w0_kernel_mod.f90` source from the
-[Part 1](../part1) directory as
+Navigate to the working directory for this part of the tutorial and open
+the supplied kernel stub file [`setval_field_w0_kernel_mod.f90`](
```
Oops, this should have been `any` kernel. Corrected.

TeranIvy commented:
Addressed the review comments and fixed broken links. Note that I also added top-level `Makefile`s to the `distributed_memory` and `single_node` directories so that `make clean` can be applied without listing every subdirectory in `tutorial/practicals/LFRic`. This now fits in nicely with the added `make clean` targets in the `distributed_memory` directory (thanks for doing that, @arporter!)

As soon as the checks have passed I'm passing back the branch.
